File: app-examples/example-jv-03adapters/src/main/java/foo/bar/example/foreadapters/ui/playlist/immutable/ImmutableListView.java
package foo.bar.example.foreadapters.ui.playlist.immutable;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.Button;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import butterknife.BindView;
import butterknife.ButterKnife;
import co.early.fore.adapters.CrossFadeRemover;
import co.early.fore.core.observer.Observer;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.ui.SyncableView;
import foo.bar.example.foreadapters.OG;
import foo.bar.example.foreadapters.R;
import foo.bar.example.foreadapters.feature.playlist.ImmutablePlaylistModel;
import foo.bar.example.foreadapters.ui.playlist.mutable.MutablePlaylistAdapter;

/**
 * Copyright © 2015-2021 early.co. All rights reserved.
 */
public class ImmutableListView extends RelativeLayout implements SyncableView {

    //models that we need
    private ImmutablePlaylistModel immutablePlaylistModel = OG.get(ImmutablePlaylistModel.class);

    //UI elements that we care about
    @BindView(R.id.playlist_totaltracks1_textview)
    public TextView totalTracksImmutable;

    @BindView(R.id.playlist_list1_recycleview)
    public RecyclerView playListImmutableRecyclerView;

    @BindView(R.id.playlist_addMany1_button)
    public Button add5ImmutableButton;

    @BindView(R.id.playlist_removeMany1_button)
    public Button clear5ImmutableButton;

    @BindView(R.id.playlist_add1_button)
    public Button addImmutableButton;

    @BindView(R.id.playlist_clear1_button)
    public Button clearImmutableButton;


    //single observer reference
    Observer observer = this::syncView;


    private ImmutablePlaylistAdapter immutablePlaylistAdapter;


    public ImmutableListView(Context context) {
        super(context);
    }

    public ImmutableListView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ImmutableListView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }


    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        ButterKnife.bind(this, this);

        setupClickListeners();

        setupAdapters();
    }

    private void setupClickListeners() {
        add5ImmutableButton.setOnClickListener(v -> immutablePlaylistModel.add5NewTracks());
        clear5ImmutableButton.setOnClickListener(v -> immutablePlaylistModel.remove5Tracks());
        addImmutableButton.setOnClickListener(v -> immutablePlaylistModel.addNewTrack());
        clearImmutableButton.setOnClickListener(v -> immutablePlaylistModel.removeAllTracks());
    }

    private void setupAdapters() {

        immutablePlaylistAdapter = new ImmutablePlaylistAdapter(immutablePlaylistModel);

        LinearLayoutManager linearLayoutManager1 = new LinearLayoutManager(getContext());
        linearLayoutManager1.setOrientation(RecyclerView.VERTICAL);

        playListImmutableRecyclerView.setLayoutManager(linearLayoutManager1);
        playListImmutableRecyclerView.setAdapter(immutablePlaylistAdapter);
        playListImmutableRecyclerView.setItemAnimator(new CrossFadeRemover());
        playListImmutableRecyclerView.setHasFixedSize(true);
    }

    //data binding stuff below

    public void syncView() {

        clear5ImmutableButton.setEnabled(immutablePlaylistModel.getItemCount() > 4);
        clearImmutableButton.setEnabled(immutablePlaylistModel.getItemCount() > 0);
        totalTracksImmutable.setText("[" + immutablePlaylistModel.getItemCount() + "]");
        immutablePlaylistAdapter.notifyDataSetChangedAuto();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        immutablePlaylistModel.addObserver(observer);
        syncView();  // <- don't forget this
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        immutablePlaylistModel.removeObserver(observer);
    }
}

File: app-examples/example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/ui/playlist/immutable/ImmutableListView.kt
package foo.bar.example.foreadapterskt.ui.playlist.immutable

import android.content.Context
import android.util.AttributeSet
import android.widget.Button
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import co.early.fore.adapters.CrossFadeRemover
import co.early.fore.core.observer.Observer
import co.early.fore.core.ui.SyncableView
import co.early.fore.kt.core.logging.Logger
import foo.bar.example.foreadapterskt.OG
import foo.bar.example.foreadapterskt.R
import foo.bar.example.foreadapterskt.feature.playlist.immutable.ImmutablePlaylistModel

/**
 * Copyright © 2015-2021 early.co. All rights reserved.
 */
class ImmutableListView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr), SyncableView {

    //models that we need to sync with
    private val immutablePlaylistModel: ImmutablePlaylistModel = OG[ImmutablePlaylistModel::class.java]
    private val logger: Logger = OG[Logger::class.java]

    private lateinit var diffableAddButton: Button
    private lateinit var diffableAdd5Button: Button
    private lateinit var diffableAdd100Button: Button
    private lateinit var diffableRemove5Button: Button
    private lateinit var diffableRemove100Button: Button
    private lateinit var diffableClearButton: Button
    private lateinit var diffableTotaltracksTextview: TextView
    private lateinit var diffableListRecycleview: RecyclerView

    private lateinit var immutablePlaylistAdapter : ImmutablePlaylistAdapter

    //single observer reference
    private var observer = Observer { syncView() }

    override fun onFinishInflate() {
        super.onFinishInflate()

        setupUiReferences()

        setupClickListeners()

        setupAdapters()
    }

    private fun setupUiReferences() {
        diffableAddButton = findViewById(R.id.diffable_add_button)
        diffableAdd5Button = findViewById(R.id.diffable_add5_button)
        diffableAdd100Button = findViewById(R.id.diffable_add100_button)
        diffableRemove5Button = findViewById(R.id.diffable_remove5_button)
        diffableRemove100Button = findViewById(R.id.diffable_remove100_button)
        diffableClearButton = findViewById(R.id.diffable_clear_button)
        diffableTotaltracksTextview = findViewById(R.id.diffable_totaltracks_textview)
        diffableListRecycleview = findViewById(R.id.diffable_list_recycleview)
    }

    private fun setupClickListeners() {
        diffableAddButton.setOnClickListener { immutablePlaylistModel.addNTracks(1) }
        diffableAdd5Button.setOnClickListener { immutablePlaylistModel.addNTracks(5) }
        diffableAdd100Button.setOnClickListener { immutablePlaylistModel.addNTracks(100) }
        diffableRemove5Button.setOnClickListener { immutablePlaylistModel.removeNTracks(5) }
        diffableRemove100Button.setOnClickListener { immutablePlaylistModel.removeNTracks(100) }
        diffableClearButton.setOnClickListener { immutablePlaylistModel.removeAllTracks() }
    }

    private fun setupAdapters() {
        immutablePlaylistAdapter = ImmutablePlaylistAdapter(immutablePlaylistModel)
        diffableListRecycleview.apply{
            layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
            adapter = immutablePlaylistAdapter
            itemAnimator = CrossFadeRemover()
            setHasFixedSize(true)
        }
    }

    override fun syncView() {
        logger.i("syncView()")
        diffableRemove5Button.isEnabled = immutablePlaylistModel.hasAtLeastNItems(5)
        diffableRemove100Button.isEnabled = immutablePlaylistModel.hasAtLeastNItems(100)
        diffableTotaltracksTextview.text = immutablePlaylistModel.getItemCount().toString()
        diffableClearButton.isEnabled = !immutablePlaylistModel.isEmpty()

        immutablePlaylistAdapter.notifyDataSetChangedAuto()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        immutablePlaylistModel.addObserver(observer)
        syncView()  // <- don't forget this
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        immutablePlaylistModel.removeObserver(observer)
    }
}
